packages/eslint-plugin/src/rules/no-unsafe-type-assertion.ts

```diff
@@ -0,0 +1,62 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import {
+  createRule,
+  getConstrainedTypeAtLocation,
+  getParserServices,
+} from '../util';
+
+export default createRule({
+  name: 'no-unsafe-type-assertion',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow type assertions that widen a type',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      unsafeTypeAssertion:
+        'Unsafe type assertion: type `{{type}}` is not assignable to type `{{asserted}}`',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function checkExpression(
+      node: TSESTree.TSAsExpression | TSESTree.TSTypeAssertion,
+    ): void {
+      const nodeType = getConstrainedTypeAtLocation(services, node.expression);
+      const assertedType = getConstrainedTypeAtLocation(
+        services,
+        node.typeAnnotation,
+      );
+
+      const isAssertionSafe = checker.isTypeAssignableTo(
+        nodeType,
+        assertedType,
+      );
+
+      if (!isAssertionSafe) {
+        context.report({
+          node,
+          messageId: 'unsafeTypeAssertion',
+          data: {
+            type: checker.typeToString(nodeType),
+            asserted: checker.typeToString(assertedType),
```

Comment:
I used `checker.typeToString()` over `getTypeName()` because with `getTypeName()` string literals get transformed into `string`. This can cause errors like ``type `string` is not assignable to type `string`.`` if a string is casted to a string literal.